File: day7/solve.py
import re
from collections import defaultdict

bottom_up = defaultdict(list)
top_down  = defaultdict(list)

with open('input.txt') as f:
    for line in f:
        outer, rhs = line.split(' bags contain ')
        for count, inner in re.findall(r'(\d) (.+?) bag', rhs):
            bottom_up[inner].append(outer)
            top_down[outer].append((int(count), inner))

def possible_outers(bottom_up, bag, seen=set()):
    for outer in bottom_up[bag]:
        if outer not in seen:
            seen.add(outer)
            possible_outers(bottom_up, outer, seen)
    return len(seen)

def total_inners(top_down, bag):
    total = 0
    for count, inner in top_down[bag]:
        total += count
        total += count * total_inners(top_down, inner)
    return total

print(possible_outers(bottom_up, 'shiny gold'))
print(total_inners(top_down, 'shiny gold'))

File: day7/solve.rb
bottom_up = Hash.new { |h, k| h[k] = [] }
top_down  = Hash.new { |h, k| h[k] = [] }

File.foreach('input.txt') do |line|
  outer, rhs = line.split(' bags contain ')
  rhs.scan /(\d) (.+?) bag/ do |count, inner|
    bottom_up[inner] << outer
    top_down[outer] << [count.to_i, inner]
  end
end

def possible_outers(bottom_up, bag, seen = [])
  bottom_up[bag].each do |outer|
    if !seen.include?(outer)
      seen << outer
      possible_outers(bottom_up, outer, seen)
    end
  end
  seen.size
end

def total_inners(top_down, bag)
  top_down[bag].sum do |count, inner|
    count * (1 + total_inners(top_down, inner))
  end
end

puts possible_outers(bottom_up, 'shiny gold')
puts total_inners(top_down, 'shiny gold')
